File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/vad/viewmodel/BackendComparisonViewModel.kt
package com.musicmuni.voxatrace.demo.sections.vad.viewmodel

import android.content.Context
import androidx.compose.ui.graphics.Color
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.calibra.CalibraVAD
import com.musicmuni.voxatrace.calibra.model.VADBackend
import com.musicmuni.voxatrace.calibra.model.VADModelProvider
import com.musicmuni.voxatrace.calibra.model.VoiceActivityLevel
import com.musicmuni.voxatrace.demo.sections.vad.model.VADBackendInfo
import com.musicmuni.voxatrace.demo.sections.vad.model.VADConstants
import com.musicmuni.voxatrace.demo.sections.vad.model.WaveformSample
import com.musicmuni.voxatrace.sonix.SonixRecorder
import com.musicmuni.voxatrace.sonix.SonixRecorderConfig
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * ViewModel for comparing two VAD backends side-by-side.
 *
 * ## VoxaTrace Integration
 * ```kotlin
 * // 1. Create VADs for two backends (auto-loads bundled models)
 * val vadLeft = CalibraVAD.create(VADModelProvider.singingRealtime())
 * val vadRight = CalibraVAD.create(VADModelProvider.speech())
 *
 * // 2. Process same audio through both
 * val ratioLeft = vadLeft.getVADRatio(samples16k)
 * val ratioRight = vadRight.getVADRatio(samples16k)
 * ```
 */
class BackendComparisonViewModel : ViewModel() {

    // Configuration
    private val _leftBackendIndex = MutableStateFlow(2) // Singing RT
    val leftBackendIndex: StateFlow<Int> = _leftBackendIndex.asStateFlow()

    private val _rightBackendIndex = MutableStateFlow(0) // Speech
    val rightBackendIndex: StateFlow<Int> = _rightBackendIndex.asStateFlow()

    // Published State - Left
    private val _vadRatioLeft = MutableStateFlow(0f)
    val vadRatioLeft: StateFlow<Float> = _vadRatioLeft.asStateFlow()

    private val _activityLevelLeft = MutableStateFlow(VoiceActivityLevel.NONE)
    val activityLevelLeft: StateFlow<VoiceActivityLevel> = _activityLevelLeft.asStateFlow()

    private val _waveformSamplesLeft = MutableStateFlow<List<WaveformSample>>(emptyList())
    val waveformSamplesLeft: StateFlow<List<WaveformSample>> = _waveformSamplesLeft.asStateFlow()

    private val _latencyMsLeft = MutableStateFlow(0)
    val latencyMsLeft: StateFlow<Int> = _latencyMsLeft.asStateFlow()

    // Published State - Right
    private val _vadRatioRight = MutableStateFlow(0f)
    val vadRatioRight: StateFlow<Float> = _vadRatioRight.asStateFlow()

    private val _activityLevelRight = MutableStateFlow(VoiceActivityLevel.NONE)
    val activityLevelRight: StateFlow<VoiceActivityLevel> = _activityLevelRight.asStateFlow()

    private val _waveformSamplesRight = MutableStateFlow<List<WaveformSample>>(emptyList())
    val waveformSamplesRight: StateFlow<List<WaveformSample>> = _waveformSamplesRight.asStateFlow()

    private val _latencyMsRight = MutableStateFlow(0)
    val latencyMsRight: StateFlow<Int> = _latencyMsRight.asStateFlow()

    // Published State - Recording
    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    // Configuration Data
    val backends = VADBackendInfo.all
    val maxWaveformSamples = 100

    // Private
    private var vadLeft: CalibraVAD? = null
    private var vadRight: CalibraVAD? = null
    private var recorder: SonixRecorder? = null
    private var recordingJob: Job? = null

    // Actions

    fun setLeftBackendIndex(index: Int) {
        val wasRecording = _isRecording.value
        if (wasRecording) {
            stopRecordingInternal()
        }
        _leftBackendIndex.value = index
        recreateVADs()
        if (wasRecording) {
            // Can't restart without context
        }
    }

    fun setRightBackendIndex(index: Int) {
        val wasRecording = _isRecording.value
        if (wasRecording) {
            stopRecordingInternal()
        }
        _rightBackendIndex.value = index
        recreateVADs()
        if (wasRecording) {
            // Can't restart without context
        }
    }

    fun toggleRecording(context: Context) {
        if (_isRecording.value) {
            stopRecording()
        } else {
            startRecording(context)
        }
    }

    fun onDisappear() {
        stopRecordingInternal()
        cleanup()
    }

    // Helpers

    fun colorForLevel(level: VoiceActivityLevel): Color {
        return when (level) {
            VoiceActivityLevel.NONE -> Color.Gray
            VoiceActivityLevel.PARTIAL -> Color(0xFFFF9800)
            VoiceActivityLevel.FULL -> Color(0xFF4CAF50)
        }
    }

    fun textForLevel(level: VoiceActivityLevel): String {
        return when (level) {
            VoiceActivityLevel.NONE -> "None"
            VoiceActivityLevel.PARTIAL -> "Partial"
            VoiceActivityLevel.FULL -> "Full"
        }
    }

    // Private Methods

    private fun createVAD(backend: VADBackend): CalibraVAD {
        return when (backend) {
            VADBackend.GENERAL -> CalibraVAD.create(VADModelProvider.general)
            VADBackend.SPEECH -> CalibraVAD.create(VADModelProvider.speech())
            VADBackend.SINGING_REALTIME -> CalibraVAD.create(VADModelProvider.singingRealtime())
        }
    }

    private fun recreateVADs() {
        vadLeft?.release()
        vadRight?.release()

        vadLeft = createVAD(backends[_leftBackendIndex.value].backend)
        vadRight = createVAD(backends[_rightBackendIndex.value].backend)

        resetStatistics()
    }

    private fun resetStatistics() {
        _vadRatioLeft.value = 0f
        _vadRatioRight.value = 0f
        _activityLevelLeft.value = VoiceActivityLevel.NONE
        _activityLevelRight.value = VoiceActivityLevel.NONE
        _waveformSamplesLeft.value = emptyList()
        _waveformSamplesRight.value = emptyList()
        _latencyMsLeft.value = 0
        _latencyMsRight.value = 0
    }

    private fun startRecording(context: Context) {
        if (vadLeft == null) {
            vadLeft = createVAD(backends[_leftBackendIndex.value].backend)
        }
        if (vadRight == null) {
            vadRight = createVAD(backends[_rightBackendIndex.value].backend)
        }

        val recordPath = "${context.cacheDir}/vad_compare_temp.m4a"
        recorder = SonixRecorder.create(recordPath, SonixRecorderConfig.VOICE)

        val rec = recorder ?: return
        val vL = vadLeft ?: return
        val vR = vadRight ?: return

        _isRecording.value = true
        resetStatistics()

        rec.start()

        recordingJob = viewModelScope.launch {
            rec.audioBuffers.collect { buffer ->
                // VOICE preset records at 16kHz; CalibraVAD handles resampling internally (ADR-017)
                val samples = buffer.samples

                // Process left
                val startLeft = System.nanoTime()
                val ratioLeft = vL.getVADRatio(samples, 16000)
                val endLeft = System.nanoTime()
                val latencyLeft = ((endLeft - startLeft) / 1_000_000).toInt()

                // Process right
                val startRight = System.nanoTime()
                val ratioRight = vR.getVADRatio(samples, 16000)
                val endRight = System.nanoTime()
                val latencyRight = ((endRight - startRight) / 1_000_000).toInt()

                if (ratioLeft < 0 && ratioRight < 0) {
                    return@collect
                }

                val amplitude = calculateRMS(samples)

                // Update left
                if (ratioLeft >= 0) {
                    _vadRatioLeft.value = ratioLeft
                    _activityLevelLeft.value = classifyLevel(ratioLeft)
                    _latencyMsLeft.value = latencyLeft

                    val sampleLeft = WaveformSample(amplitude, ratioLeft > 0.5f)
                    val newSamplesLeft = _waveformSamplesLeft.value.toMutableList()
                    newSamplesLeft.add(sampleLeft)
                    if (newSamplesLeft.size > maxWaveformSamples) {
                        newSamplesLeft.removeAt(0)
                    }
                    _waveformSamplesLeft.value = newSamplesLeft
                }

                // Update right
                if (ratioRight >= 0) {
                    _vadRatioRight.value = ratioRight
                    _activityLevelRight.value = classifyLevel(ratioRight)
                    _latencyMsRight.value = latencyRight

                    val sampleRight = WaveformSample(amplitude, ratioRight > 0.5f)
                    val newSamplesRight = _waveformSamplesRight.value.toMutableList()
                    newSamplesRight.add(sampleRight)
                    if (newSamplesRight.size > maxWaveformSamples) {
                        newSamplesRight.removeAt(0)
                    }
                    _waveformSamplesRight.value = newSamplesRight
                }
            }
        }
    }

    private fun stopRecordingInternal() {
        recordingJob?.cancel()
        recorder?.stop()
        _isRecording.value = false
        vadLeft?.reset()
        vadRight?.reset()
    }

    fun stopRecording() {
        stopRecordingInternal()
    }

    private fun cleanup() {
        recorder?.release()
        recorder = null
        vadLeft?.release()
        vadLeft = null
        vadRight?.release()
        vadRight = null
    }

    private fun calculateRMS(samples: FloatArray): Float {
        if (samples.isEmpty()) return 0f
        val sumSquares = samples.fold(0f) { acc, sample -> acc + sample * sample }
        return sqrt(sumSquares / samples.size)
    }

    private fun classifyLevel(ratio: Float): VoiceActivityLevel {
        return when {
            ratio < VADConstants.noSingingThreshold -> VoiceActivityLevel.NONE
            ratio < VADConstants.partialSingingThreshold -> VoiceActivityLevel.PARTIAL
            else -> VoiceActivityLevel.FULL
        }
    }

    override fun onCleared() {
        super.onCleared()
        cleanup()
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/VAD/ViewModel/BackendComparisonViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for comparing two VAD backends side-by-side.
///
/// ## VoxaTrace Integration (~10 lines)
/// ```swift
/// // 1. Create VADs for two backends (auto-loads bundled models)
/// let vadLeft = CalibraVAD.create(.singingRealtime())
/// let vadRight = CalibraVAD.create(.speech())
///
/// // 2. Process same audio through both
/// let ratioLeft = vadLeft.getVADRatio(samples: samples16k)
/// let ratioRight = vadRight.getVADRatio(samples: samples16k)
/// ```
@MainActor
final class BackendComparisonViewModel: ObservableObject {

    // MARK: - Configuration

    @Published var leftBackendIndex: Int = 2 {
        didSet { recreateVADs() }
    }
    @Published var rightBackendIndex: Int = 0 {
        didSet { recreateVADs() }
    }

    // MARK: - Published State - Left

    @Published private(set) var vadRatioLeft: Float = 0.0
    @Published private(set) var activityLevelLeft: VoiceActivityLevel = .none
    @Published private(set) var waveformSamplesLeft: [WaveformSample] = []
    @Published private(set) var latencyMsLeft: Int = 0

    // MARK: - Published State - Right

    @Published private(set) var vadRatioRight: Float = 0.0
    @Published private(set) var activityLevelRight: VoiceActivityLevel = .none
    @Published private(set) var waveformSamplesRight: [WaveformSample] = []
    @Published private(set) var latencyMsRight: Int = 0

    // MARK: - Published State - Recording

    @Published private(set) var isRecording = false

    // MARK: - Configuration Data

    let backends = VADBackendInfo.all
    let maxWaveformSamples = 100

    // MARK: - Private

    private var vadLeft: CalibraVAD?
    private var vadRight: CalibraVAD?
    private var recorder: SonixRecorder?

    // MARK: - Lifecycle

    func onDisappear() {
        stopRecording()
        cleanup()
    }

    // MARK: - Actions

    func toggleRecording() {
        if isRecording {
            stopRecording()
        } else {
            startRecording()
        }
    }

    // MARK: - Helpers

    func colorForLevel(_ level: VoiceActivityLevel) -> Color {
        switch level {
        case .none: return .gray
        case .partial: return .orange
        case .full: return .green
        @unknown default: return .gray
        }
    }

    func textForLevel(_ level: VoiceActivityLevel) -> String {
        switch level {
        case .none: return "None"
        case .partial: return "Partial"
        case .full: return "Full"
        @unknown default: return "Unknown"
        }
    }

    // MARK: - Private Methods

    private func createVAD(for backend: VADBackend) -> CalibraVAD {
        switch backend {
        case .general:
            return CalibraVAD.create(.general)
        case .speech:
            // Auto-loads bundled Silero model
            return CalibraVAD.create(.speech())
        case .singingRealtime:
            // Auto-loads bundled SwiftF0 model
            return CalibraVAD.create(.singingRealtime())
        default:
            return CalibraVAD.create(.general)
        }
    }

    private func recreateVADs() {
        let wasRecording = isRecording
        if wasRecording {
            stopRecording()
        }

        vadLeft?.release()
        vadRight?.release()

        vadLeft = createVAD(for: backends[leftBackendIndex].backend)
        vadRight = createVAD(for: backends[rightBackendIndex].backend)

        resetStatistics()

        if wasRecording {
            startRecording()
        }
    }

    private func resetStatistics() {
        vadRatioLeft = 0.0
        vadRatioRight = 0.0
        activityLevelLeft = .none
        activityLevelRight = .none
        waveformSamplesLeft.removeAll()
        waveformSamplesRight.removeAll()
        latencyMsLeft = 0
        latencyMsRight = 0
    }

    private func startRecording() {
        if vadLeft == nil {
            vadLeft = createVAD(for: backends[leftBackendIndex].backend)
        }
        if vadRight == nil {
            vadRight = createVAD(for: backends[rightBackendIndex].backend)
        }

        let tempPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("vad_compare_temp.m4a").path
        recorder = SonixRecorder.create(outputPath: tempPath, config: .voice, audioSession: .recording)

        guard let recorder = recorder,
              let vadLeft = vadLeft,
              let vadRight = vadRight else { return }

        isRecording = true
        resetStatistics()

        // ADR-017: Pass sampleRate directly; CalibraVAD handles resampling internally
        Task {
            let hwRate = AudioSessionManager.hardwareSampleRate

            for await buffer in recorder.audioBuffers {
                let startLeft = CFAbsoluteTimeGetCurrent()
                let ratioLeft = vadLeft.getVADRatio(samples: buffer.samples, sampleRate: hwRate)
                let endLeft = CFAbsoluteTimeGetCurrent()
                let latencyLeft = Int((endLeft - startLeft) * 1000)

                let startRight = CFAbsoluteTimeGetCurrent()
                let ratioRight = vadRight.getVADRatio(samples: buffer.samples, sampleRate: hwRate)
                let endRight = CFAbsoluteTimeGetCurrent()
                let latencyRight = Int((endRight - startRight) * 1000)

                if ratioLeft < 0 && ratioRight < 0 {
                    continue
                }

                let amplitude = calculateRMS(samples: buffer.samples)

                await MainActor.run {
                    if ratioLeft >= 0 {
                        self.vadRatioLeft = ratioLeft
                        self.activityLevelLeft = self.classifyLevel(ratio: ratioLeft)
                        self.latencyMsLeft = latencyLeft

                        let sampleLeft = WaveformSample(amplitude: amplitude, isVoice: ratioLeft > 0.5)
                        self.waveformSamplesLeft.append(sampleLeft)
                        if self.waveformSamplesLeft.count > self.maxWaveformSamples {
                            self.waveformSamplesLeft.removeFirst()
                        }
                    }

                    if ratioRight >= 0 {
                        self.vadRatioRight = ratioRight
                        self.activityLevelRight = self.classifyLevel(ratio: ratioRight)
                        self.latencyMsRight = latencyRight

                        let sampleRight = WaveformSample(amplitude: amplitude, isVoice: ratioRight > 0.5)
                        self.waveformSamplesRight.append(sampleRight)
                        if self.waveformSamplesRight.count > self.maxWaveformSamples {
                            self.waveformSamplesRight.removeFirst()
                        }
                    }
                }
            }
        }

        recorder.start()
    }

    private func stopRecording() {
        recorder?.stop()
        isRecording = false
        vadLeft?.reset()
        vadRight?.reset()
    }

    private func cleanup() {
        recorder?.stop()
        recorder?.release()
        recorder = nil

        vadLeft?.release()
        vadLeft = nil

        vadRight?.release()
        vadRight = nil
    }

    private func calculateRMS(samples: [Float]) -> Float {
        guard !samples.isEmpty else { return 0 }
        let sumSquares = samples.reduce(0) { $0 + $1 * $1 }
        return sqrt(sumSquares / Float(samples.count))
    }

    private func classifyLevel(ratio: Float) -> VoiceActivityLevel {
        if ratio < VADConstants.noSingingThreshold {
            return .none
        } else if ratio < VADConstants.partialSingingThreshold {
            return .partial
        } else {
            return .full
        }
    }
}

import SwiftUI

extension BackendComparisonViewModel {
    // SwiftUI Color helper that needs to be in an extension with SwiftUI import
}
